Prefer connecting to dev server when running in DEBUG

peppy opened the pull request:
Minimum implementation, meant for visibility. Needs further discussion on how to provide this configuration (probably not via #if and probably all in one place).

This server will require registering a new account (from within lazer). It will allow playing / testing multiplayer and everything else, but some features like beatmap download will fail due to a limited deployment.

peppy committed:
Prefer connecting to dev server when running in DEBUG

bdach commented:
I'm somewhat certain that the typical .NET-Core-ish way to do this is an `appsettings.json`. Would probably be something like:

```json
{
	"api": {
		"endpoint": "https://dev.ppy.sh",
		"client_secret": "3LP2mhUrV89xxzD1YKNndXHEhWWCRLPNKioZ9ymT",
		"client_id": 5
	},
	"spectator": {
		"endpoint": "https://dev.ppy.sh/spectator"
	},
	"multiplayer": {
		"endpoint": "https://dev.ppy.sh/multiplayer"
	}
}
```

Of course there would be a corresponding `appsettings.Release.json`. Not sure how to jam that into our setup yet as we don't have the typical DI ASP.NET Core does, but throwing this out as a format suggestion first and foremost before digging further in that direction.

bdach commented:
On checking there is a Microsoft-endorsed way to use that format for console apps, but it [somewhat hijacks `Program.Main`](https://docs.microsoft.com/en-us/dotnet/core/extensions/configuration)... Not sure whether we want to go for that in that case?

bdach commented:
Here's a rough draft branch: https://github.com/ppy/osu/compare/master...bdach:dev-server-config-working

Can be improved in places, generally meant as a MVP. Seems to launch fine in both configs; API doesn't work in debug for what I presume to be obvious reasons (different user set?)

ThePooN commented:
Yes, it's a completely isolated environment. You need to create a seperate account via lazer.

bdach commented:
After some discussion in discord I think I'd agree with the sentiment expressed therein of not forcing this always for debug, but instead having alternative run configs for dev/production osu-web instances, just so you can switch on a dime. I wouldn't mind all that much if the two instances were full parity, but they aren't, and i.e. debugging potential osu!direct download issues on release builds could end up being a pain.

Both this PR in its current state and my draft thing can be adapted to that model by just choosing a different arbitrary compilation symbol to switch to, I guess.

peppy commented:
The intended direction is to always use this for debug. Anything missing can be added.

peppy committed:
Move all endpoint information to a configuration class

peppy committed:
Add website root URL and update most links to use it

For what it's worth, I intentionally didn't include news / changelog /
supporter, because these should never change.

peppy committed:
Fix various tests failing due to dependence on specific online data

bdach committed:
Trim stray newline

bdach committed:
Remove explicit public access modifier from interface

bdach commented on osu.Game/Online/API/Requests/Responses/APIChangelogEntry.cs:
```diff
@@ -48,6 +48,7 @@ public class APIChangelogEntry
         Add,
-        Fix
+        Fix,
+        Misc
```
sort of a drive-by change, but i'll look away for that one

peppy replied:
This is actually a valid type, just haven't used it in years.

bdach replied:
Yeah I checked web, just saying that it's not hugely related to the main purpose of the PR.

bdach commented on osu.Game/Overlays/Changelog/ChangelogBuild.cs:
```diff
@@ -131,33 +131,36 @@ private void load(OsuColour colours, OverlayColourProvider colourProvider)
+                                Id = entry.GithubUser.UserId.Value
+                            }, t =>
+                            {
+                                t.Font = fontMedium;
+                                t.Colour = entryColour;
+                            });
+                        }
+                        else if (entry.GithubUser.GithubUrl != null)
                         {
-                            t.Font = fontMedium;
-                            t.Colour = entryColour;
-                        });
-                    }
-                    else
-                    {
-                        title.AddText(entry.GithubUser.DisplayName, t =>
+                            title.AddLink(entry.GithubUser.DisplayName, entry.GithubUser.GithubUrl, t =>
+                            {
+                                t.Font = fontMedium;
+                                t.Colour = entryColour;
+                            });
+                        }
+                        else
                         {
-                            t.Font = fontMedium;
-                            t.Colour = entryColour;
-                        });
+                            title.AddText(entry.GithubUser.DisplayName, t =>
+                            {
+                                t.Font = fontMedium;
+                                t.Colour = entryColour;
+                            });
+                        }
                     }
```
This is super ugly but it wasn't pretty before, and refactoring this isn't in scope of this PR, so again, probably going to look away but open an issue for this and the disabled test.
